lxd/network/openvswitch/docs/developer-documentation.md

```diff
@@ -0,0 +1,127 @@
+> If no members have the role, all cluster members will be added to the chassis group.
+
+### OVS (OpenVSwitch) Bridge
+OVS bridges are used to connect virtual networks to physical ones and vice-versa.
+In LXD, if OpenVSwitch APIs are invoked directly, that means changes are being applied on the same physical cluster member.
+
+For each LXD cluster member there are two OVS bridges:
+1. The uplink bridge - this connects the uplink network on the host to the external switch inside each OVN network.
+2. The integration bridge - this connects instances to the internal switch inside each OVN network.
+
+### Chassis group
+A chassis group is an indirection between physical chassis and the virtual networks that use them.
+Each LXD OVN network has one chassis group.
+This allows us to, for example, set chassis priority on a per-network basis so that not all ingress/egress is happening on a single cluster member.
+
+### OVN underlay
+The OVN underlay is the means by which networks are virtualized across cluster members.
+It is a Geneve tunnel which creates a layer 2 overlay network across layer 3 infrastructure.
+The OVN underlay is configured and managed by OVN.
+
+### Logical router
+A logical router is a virtualized router.
+There is one per LXD OVN network.
+This handles L3 networking and additionally has associated NAT rules and security policies.
+
+### Logical switch
+Logical routers cannot be directly connected to OVS bridges; for this we use a logical switch.
+There are two logical switches per LXD OVN network:
+- The external switch, which is between the logical router and the chassis group (which indirects the to uplink bridge, which may be on another cluster member).
+- The internal switch, which is between the logical router and the integration bridge. This switch contains DHCP and IP allocation configuration.
+
+### Logical switch/router ports
+In OVN, when you create a router or switch it doesn't initially have any ports.
+You need to create ports and then link them.
+For example, the internal logical switch and the logical router for a LXD OVN network are connected by:
+1. Adding a logical router port to the logical router.
+2. Adding a logical switch port to the internal logical switch.
+3. Configuring the internal logical switch port as a router port and setting the logical router name.
+
+Some configuration is applied directly at port level.
+For example, in a LXD OVN network, IPv6 router advertisement settings are applied on the logical router port for the internal switch.
+This is by design. It allows OVN to push configuration down to the port level so that packets are handled as quickly as possible.
+
+### Port groups
+When an instance is attached via the integration bridge to the internal switch of a LXD OVN network, the logical switch port that is created is added to the port group for that network.
+This is used for ACL configuration so that rules can be applied on the whole address set rather than individual ports.
```

Comment:
whole address? whole network right?

Reply:
whole address *set*, but probably network is more accurate.

Reply:
whole address **set**, but probably network is more accurate.

Reply:
I got this terminology from the OVN northbound DB schema docs. A port group can be used in a `match` rule for the ACL table, as can address sets. When a port group is created or modified, entries in the `Address_Set` table are automatically populated. So using `Port_Groups` simplifies management of ACLs.

I struggled to put that into words :laughing:

Reply:
I suggest preceding this sentence with a brief description of the phrase "address set" as used in OVN.

Reply:
ah i missed the "set" parse as i skimmed it.